Expiration period for logged-in users MLN-231

LucasCioffi opened the pull request.

nnat425 commented on .travis.yml:
```diff
@@ -26,7 +26,7 @@ deploy:
   on:
     repo: NYPL/MyLibraryNYCApp
-    branch: development
+    branch: lc_timeout3
```
REMINDER: Change back to development.

nnat425 commented on app/assets/javascripts/angular.app.js.erb:
```diff
@@ -57,14 +57,14 @@ app.run(['Angularytics', '$http', function(Angularytics, $http) {
-    // Analytics maintinas these internal names for these custom dimensions:
+  $http.get('/settings.json').success(function(settings) {
+
+    // Analytics maintains these internal names for these custom dimensions:
```
Didn't you get rid of this information in another PR?

LucasCioffi replied:
@nnat425 Yes, but that one may not be merged in soon, because it's not a priority to finish.

nnat425 commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,39 @@
+  layout 'minimal', :only => [ :timeout_check ]
+
+  def timeout_check
+    return if !session || !session["warden.user.user.session"]
```
What value does warden.user.user.session store? Is it a timestamp? Also can we change the name of the key to something more clear?

LucasCioffi replied:
Sounds good; I added this:
```
# session["warden.user.user.session"] comes from adding timeoutable to the user model
# example: {"last_request_at"=>1537283353}
```

nnat425 commented on app/views/layouts/_timeout.html:
```diff
@@ -0,0 +1,18 @@
+  function PeriodicalQuery() {
+    var request = new XMLHttpRequest();
+    request.onload = function (event) {
+      console.log(request.response['timeout_status']); // turn this on for debugging locally
```
Comment out?

LucasCioffi replied:
It's now deleted.

nnat425 commented on app/views/layouts/minimal.html.erb:
```diff
@@ -0,0 +1 @@
+<!-- Intentionally left blank -->
```
Maybe explain why this is intentionally left blank?

LucasCioffi replied:
It's now renamed to "Empty".  Sometimes we would prefer an empty layout rather than the application layout.
Sessions Controller: `layout 'empty', :only => [ :timeout_check ]`
Home Controller: `layout 'empty', :only => [ :extend_session_iframe ]`

The application layout has a bunch of files being required such as `<%= render "layouts/timeout" if current_user %>` that we do not need when performing this XMLHttpRequest.

nnat425 commented on config/routes.rb:
```diff
@@ -1,14 +1,21 @@
+    match 'timeout' => 'sessions#timeout', via: :get
+  end
+
+  get 'extend_session_iframe' => 'home#extend_session_iframe'
```
Maybe add some comments on this route? A future developer might be caught off guard by this

LucasCioffi replied:
I added a comment in the controller rather than the routes.rb.

LucasCioffi commented:
@nnat425 I addressed your comments.  In the most recent commit, I documented the methods you said might need it.

I also set it up for Chris to QA in the Development environment (with browser console debug messages and 30-second timeout).  Prior to merging to the Development branch, I'll reset the times to the following:
```
In _timeout.html, change this:
  setTimeout(PeriodicalQuery, (3 * 1000)); # for checking every 3 seconds
to this:
  setTimeout(PeriodicalQuery, (5 * 1000 * 60)); # for checking every 5 minutes

In the max_session_duration method in sessions_controller.rb, change this:
  8.hours
to this:
  15.seconds

In the timeout_warning_duration method in sessions_controller.rb, change this:
  20.seconds
to this:
  10.minutes
```

nnat425 commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,35 @@
+  layout 'empty', :only => [ :timeout_check ]
+
+  def timeout_check
+    return if !session || !session["warden.user.user.session"] # session["warden.user.user.session"] comes from adding timeoutable to the user model
```
What value is stored "warden.user.user.session". Can we change from the name of this key?

LucasCioffi replied:
It's the default from Warden.  I'm thinking we should keep it the same, because a bunch of things may depend upon it now or in the future.

LucasCioffi replied:
Ok I added this to the `timeout_check` method in sessions_controller.rb:
```
# session["warden.user.user.session"] comes from adding timeoutable to the user model
# example: {"last_request_at"=>1537283353}
```

nnat425 commented on app/views/sessions/timeout_check.js:
```diff
@@ -0,0 +1 @@
+// this gets called if there's no session
```
Why are we calling an empty javascript file?

nnat425 replied:
I understand it's being called because there is no session, so why not call anything at all if the file is empty?

daryanypl replied:
I think the doc string might be misleading.  Lucas, is is more like: "this gets called in the process of checking the user's session timeout"?

LucasCioffi replied:
@daryanypl I removed this file; the JS request to SessionsController now renders a JSON response if there's no session: `render json: { 'timeout_status': 'no_current_user' }, status: 100`

daryanypl commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,35 @@
+  end
+
+  def timeout_warning_duration
+    20.seconds
```
these look like development values

daryanypl commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,35 @@
+
+    if session["warden.user.user.session"] && Time.zone.now.to_i > timeout_timestamp.to_i
+      sign_out current_user
+      flash[:notice] = "Your session has timed out."
```
how long does the default flash notice present itself for?

LucasCioffi replied:
@daryanypl Just 5 seconds.

daryanypl commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,35 @@
+  layout 'empty', :only => [ :timeout_check ]
+
+  def timeout_check
+    return if !session || !session["warden.user.user.session"] # session["warden.user.user.session"] comes from adding timeoutable to the user model
```
Thanks, I appreciate the doc string.

daryanypl commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,35 @@
+  end
+
+  def timeout_timestamp
+    session["warden.user.user.session"]["last_request_at"] + max_session_duration
```
Do we know that warden uses the correct Time.zone.now, not the Time.now?

daryanypl commented on app/controllers/settings_controller.rb:
```diff
@@ -1,12 +1,14 @@
       session[:redirect_after_login] = "/users/edit"
       puts "format: ", request.format
-      redirect_to new_user_session_path if request.format == 'html'
+      if request.format == 'html'
```
Do we need to consider non-html formats?

daryanypl commented on app/views/layouts/_timeout.html:
```diff
@@ -0,0 +1,18 @@
+    request.open('GET', '/timeout_check', true);
+    request.responseType = 'json';
+    request.send();
+    setTimeout(PeriodicalQuery, (3 * 1000));
```
development values?

daryanypl commented on config/application.rb:
```diff
@@ -38,9 +38,9 @@ class Application < Rails::Application
-    config.exceptions_app = lambda do |env|
-      ExceptionsController.action(:render_error).call(env)
-    end
+    # config.exceptions_app = lambda do |env|
```
What's this bit?
